operator/v1/types_csi_cluster_driver.go

```diff
@@ -281,6 +281,27 @@ type VSphereCSIDriverConfigSpec struct {
 	// will be rejected.
 	// +optional
 	TopologyCategories []string `json:"topologyCategories,omitempty"`
+
+	// globalMaxSnapshotsPerBlockVolume is a global configuration parameter that applies to volumes on all kinds of
+	// datastores. By default, it is set to three.
+	// +kubebuilder:validation:Optional
+	// +openshift:enable:FeatureSets=TechPreviewNoUpgrade
+	// +optional
+	GlobalMaxSnapshotsPerBlockVolume *uint32 `json:"globalMaxSnapshotsPerBlockVolume,omitempty"`
```

Comment:
For all new fields:

* There should be some validation, for example the value should not be negative. Or semantics of negative values should be described.
* Are there any max values?
* Value `0` should have some description - is it a valid value? Does it turn off snapshot support for VSAN / VVOL / globally? 
* What are the impacts of setting the values high? I would expect higher values mean less performance, otherwise we can set the values to MAXINT by default and we don't need any configuration.
* It could help if there was link to vSphere docs, the answers could be answered there.

Reply:
> There should be some validation, for example the value should not be negative. Or semantics of negative values should be described.

- Correct - currently the operator would fail to unmarshal as the field is uint32 type
- Adding validation for min/max volumes.

```
W0312 17:09:26.526662    8935 reflector.go:539] k8s.io/client-go@v0.29.1/tools/cache/reflector.go:229: failed to list *v1.ClusterCSIDriver: json: cannot unmarshal number -1 into Go struct field VSphereCSIDriverConfigSpec.items.spec.driverConfig.vSphere.globalMaxSnapshotsPerBlockVolume of type uint32
E0312 17:09:26.526720    8935 reflector.go:147] k8s.io/client-go@v0.29.1/tools/cache/reflector.go:229: Failed to watch *v1.ClusterCSIDriver: failed to list *v1.ClusterCSIDriver: json: cannot unmarshal number -1 into Go struct field VSphereCSIDriverConfigSpec.items.spec.driverConfig.vSphere.globalMaxSnapshotsPerBlockVolume of type uint32
```

> Are there any max values?

- From VMware documentation: `Maximum of 32 snapshots are **supported** in a chain. However, for a better performance use only 2 to 3 snapshots.`
- Adding validation for max value of 32.
- Setting any higher value in the config is not prevented currently:
```
[Snapshot]
global-max-snapshots-per-block-volume = 33
```

> Value `0` should have some description - is it a valid value? Does it turn off snapshot support for VSAN / VVOL / globally?

- Not documented at all, after testing it seems like `0` value results in disabling the option and default is used.
- Snapshots are not disabled when `0` is set
- Setting min value to 1 should work fine as there is no reason to set `0`

```
$ oc -n openshift-cluster-csi-drivers get cm/vsphere-csi-config -o yaml | grep global-max-snapshots-per-block-volume
    global-max-snapshots-per-block-volume = 0
```

```
message: 'Failed to create snapshot: failed to take snapshot of the volume c0e2dc41-33de-4566-9323-9aa422741493: "rpc error: code = FailedPrecondition desc = the number of snapshots on the source volume c0e2dc41-33de-4566-9323-9aa422741493 reaches the configured maximum (3)"'
```

> What are the impacts of setting the values high? I would expect higher values mean less performance, otherwise we can set the values to MAXINT by default and we don't need any configuration.

- This is documented under "Best practices for snapshots": https://kb.vmware.com/s/article/1025279
- 2-3 snapshots are recommended for better performance so we can expect a performance drop when using more than 3 snapshots

 > It could help if there was link to vSphere docs, the questions could be answered there.

- Do we actually want to link to external documentation from comments here?
 - Adding the link to relevant doc: https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/3.0/vmware-vsphere-csp-getting-started/GUID-E0B41C69-7EEB-450F-A73D-5FD2FF39E891.html

Reply:
Thanks for the update. I think linking vSphere docs will help users to actually understand what the options are good for and they don't need to guess.